compiler/passes/normalize.cpp

```diff
@@ -989,52 +989,67 @@ static void normalizeReturns(FnSymbol* fn) {
   for_vector(CallExpr, ret, rets) {
     SET_LINENO(ret);
 
-    if (isIterator == true) {
-      INT_ASSERT(retval != NULL);
-
-      // Three cases:
-      // (1) yield expr; => mov _ret expr; yield _ret;
-      // (2) return; => goto end_label;
-      // (3) return expr; -> mov _ret expr; yield _ret; goto end_label;
-      // Notice how (3) is the composition of (1) and (2).
-      if (isVoidReturn(ret) == false) { // Cases 1 and 3
-        insertRetMove(fn, retval, ret);
-        ret->insertBefore(new CallExpr(PRIM_YIELD, retval));
-      }
+    if (isIterator == false && retval != NULL) {
+      insertRetMove(fn, retval, ret);
+    }
 
-      if (ret->isPrimitive(PRIM_YIELD) == true) { // Case 1 only.
-        // it's a yield => no goto; need to remove the original node
-        ret->remove();
-      } else {    // Cases 2 and 3.
-        if (ret->next != label->defPoint) {
-          ret->replace(new GotoStmt(GOTO_RETURN, label));
+    // replace with GOTO(label)
+    if (ret->next != label->defPoint) {
+      ret->replace(new GotoStmt(GOTO_RETURN, label));
 
-          labelIsUsed = true;
-        } else {
-          ret->remove();
-        }
-      }
+      labelIsUsed = true;
+    } else {
+      ret->remove();
+    }
+  }
+
+  if (labelIsUsed == false) {
+    label->defPoint->remove();
+  }
+}
 
+static void normalizeYields(FnSymbol* fn) {
+  INT_ASSERT(fn->isIterator());
+  SET_LINENO(fn);
 
-    // Not an iterator
-    } else {
-      if (retval != NULL) {
-        insertRetMove(fn, retval, ret);
-      }
+  std::vector<CallExpr*> yields;
+  std::vector<CallExpr*> calls;
 
-      // replace with GOTO(label)
-      if (ret->next != label->defPoint) {
-        ret->replace(new GotoStmt(GOTO_RETURN, label));
+  collectMyCallExprs(fn, calls, fn);
 
-        labelIsUsed = true;
-      } else {
-        ret->remove();
-      }
+  for_vector(CallExpr, call, calls) {
+    if (call->isPrimitive(PRIM_YIELD)) {
+      yields.push_back(call);
     }
   }
 
-  if (labelIsUsed == false) {
-    label->defPoint->remove();
+  // For each yield statement, adjust it similarly to a return.
+  for_vector(CallExpr, yield, yields) {
+    SET_LINENO(yield);
+
+    // Create a new YVV variable
+    // MPF: I don't think YVV or RVV need to exist in the long term,
+    // but using YVV enables minor adjustment in most of the rest of
+    // the compiler.
+    VarSymbol* retval = newTemp("yret", fn->retType);
+    retval->addFlag(FLAG_YVV);
+
+    if (fn->retTag == RET_PARAM) {
+      retval->addFlag(FLAG_PARAM);
+    }
+
+    if (fn->retTag == RET_TYPE) {
+      retval->addFlag(FLAG_TYPE_VARIABLE);
+    }
+
+    if (fn->hasFlag(FLAG_MAYBE_TYPE)) {
+      retval->addFlag(FLAG_MAYBE_TYPE);
+    }
+
```

Comment:
I'm not so sure FLAG_MAYBE_TYPE or FLAG_MAYBE_PARAM make sense, since AFAIK we can't yield a param or a type from an iterator.

>  error: iterators may not yield or return parameters
>  error: iterators may not yield or return types

So I'll remove those two.

At this point in the compiler, we're supposed to know whether we're returning a ref or not by the function's return intent. This code calls insertRetMove which adds PRIM_ADDR_OF or PRIM_DEREF depending on whether or not the function returns with ref intent.